cmd/snap-confine/selinux-support.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "config.h"
+#include "selinux-support.h"
+
+#include <selinux/restorecon.h>
+#include <selinux/selinux.h>
+#include <selinux/context.h>
+
+#include "../libsnap-confine-private/utils.h"
+#include "../libsnap-confine-private/string-utils.h"
+
+int sc_selinux_relabel_run_dir(void)
+{
+	if (is_selinux_enabled() < 1) {
+		return 0;
+	}
+
+	int ret = selinux_restorecon("/run/snapd",
+				     SELINUX_RESTORECON_RECURSE |
+				     SELINUX_RESTORECON_IGNORE_MOUNTS |
+				     SELINUX_RESTORECON_XDEV);
+	if (ret == -1) {
+		die("failed to restore context of /run/snapd");
+	}
+	return ret;
+}
+
+/**
+ * Set security context for the snap
+ *
+ **/
+int sc_selinux_set_snap_execcon(void)
+{
+	/* die("foo"); */
+	if (is_selinux_enabled() < 1) {
+		debug("selinux not enabled");
+		return 0;
+	}
+
+	char *ctx_str = NULL;
```

Comment:
Can you please add a selinux specific cleanup handler that calls `freecon`

This is based on https://linux.die.net/man/3/getcon`